lib/PuppeteerSharp/Browser.cs

```diff
@@ -220,17 +214,15 @@ void TargetHandler(object sender, TargetChangedArgs e)
 
             try
             {
-                foreach (var target in Targets())
-                {
-                    if (predicate(target))
-                    {
-                        return target;
-                    }
-                }
-
                 TargetCreated += TargetHandler;
                 TargetChanged += TargetHandler;
 
+                var existingTarget = Targets().FirstOrDefault(predicate);
+                if (existingTarget != null)
+                {
+                    return existingTarget;
+                }
```

Comment:
The issue is that in .NET two threads could execute code in parallel. If we check for `Targets()` before adding the event handlers, there's a possibility that it returns empty and then, in another thread, a `Target` is created, even before the event handlers are attached in the original thread. In this scenario, the creation won't be observed by the event listeners, and `targetCompletionSource` will never resolve. It's an improbable race condition, which is why it is not easy to reproduce and was undetected so far, but if you run enough tests it eventually happens.

In JavaScript this scenario doesn't happen, because it is single-threaded, so there's a guarantee that `Targets()` won't change between the enumeration and the event handlers being added.

Reply:
Roger, I hadn't fully understood your first explanation.